File: software/examples/javascript/ExampleCallback.js
var Tinkerforge = require('tinkerforge');

var HOST = 'localhost';
var PORT = 4223;
var UID = 'iB4'; // Change to your UID

var ipcon = new Tinkerforge.IPConnection(); // Create IP connection
var m = new Tinkerforge.BrickletMoisture(UID, ipcon); // Create device object

ipcon.connect(HOST, PORT,
    function(error) {
        console.log('Error: '+error);
    }
); // Connect to brickd
// Don't use device before ipcon is connected

ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    function(connectReason) {
        // Set Period for moisture callback to 1s (1000ms)
        // Note: The moisture callback is only called every second if the
        // moisture has changed since the last call!
        m.setMoistureCallbackPeriod(1000);
    }
);

// Register moisture callback
m.on(Tinkerforge.BrickletMoisture.CALLBACK_MOISTURE,
    // Callback function for moisture value
    function(moisture) {
        console.log('Moisture Value: '+moisture);
    }
);

console.log("Press any key to exit ...");
process.stdin.on('data',
    function(data) {
        ipcon.disconnect();
        process.exit(0);
    }
);

File: software/examples/rust/example_callback.rs
use std::{io, error::Error};
use std::thread;
use tinkerforge::{ip_connection::IpConnection, 
                  moisture_bricklet::*};


const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Moisture Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let m = MoistureBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
    // Don't use device before ipcon is connected.

     let moisture_receiver = m.get_moisture_callback_receiver();

        // Spawn thread to handle received callback messages. 
        // This thread ends when the `m` object
        // is dropped, so there is no need for manual cleanup.
        thread::spawn(move || {
            for moisture in moisture_receiver {           
                		println!("Moisture Value: {}", moisture);
            }
        });

		// Set period for moisture value receiver to 1s (1000ms).
		// Note: The moisture value callback is only called every second
		//       if the moisture value has changed since the last call!
		m.set_moisture_callback_period(1000);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}
